internal/controller/servermaintenance_controller.go

```diff
@@ -239,41 +246,29 @@ func (r *ServerMaintenanceReconciler) setAndPatchServerPowerState(ctx context.Co
 	return nil
 }
 
-func (r *ServerMaintenanceReconciler) patchServerRef(ctx context.Context, log logr.Logger, maintenance *metalv1alpha1.ServerMaintenance, server *metalv1alpha1.Server) error {
+func (r *ServerMaintenanceReconciler) updateServerRef(ctx context.Context, log logr.Logger, maintenance *metalv1alpha1.ServerMaintenance, server *metalv1alpha1.Server) error {
 	if server.Spec.ServerMaintenanceRef != nil {
 		log.V(1).Info("Server is already in Maintenance", "Server", server.Name, "Maintenance", server.Spec.ServerMaintenanceRef.Name)
 		return nil
 	}
-	if server.Spec.ServerMaintenanceRef == nil {
-		serverBase := server.DeepCopy()
-		server.Spec.ServerMaintenanceRef = &v1.ObjectReference{
-			APIVersion: "metal.ironcore.dev/v1alpha1",
-			Kind:       "ServerMaintenance",
-			Namespace:  maintenance.Namespace,
-			Name:       maintenance.Name,
-			UID:        maintenance.UID,
-		}
-		if err := r.Patch(ctx, server, client.MergeFrom(serverBase)); err != nil {
-			return fmt.Errorf("failed to patch maintenance ref for server: %w", err)
-		}
-		log.V(1).Info("Patched ServerMaintenance reference on Server", "Server", server.Name, "ServerMaintenanceeRef", maintenance.Name)
-	}
-	return nil
-}
-
-func (r *ServerMaintenanceReconciler) handleCompletedState(ctx context.Context, log logr.Logger, serverMaintenance *metalv1alpha1.ServerMaintenance) (ctrl.Result, error) {
-	server, err := r.getServerRef(ctx, serverMaintenance)
-	if err != nil {
-		return ctrl.Result{}, err
+	server.Spec.ServerMaintenanceRef = &v1.ObjectReference{
+		APIVersion: "metal.ironcore.dev/v1alpha1",
+		Kind:       "ServerMaintenance",
+		Namespace:  maintenance.Namespace,
+		Name:       maintenance.Name,
+		UID:        maintenance.UID,
 	}
-	log.V(1).Info("Server maintenance completed", "Server", server.Name)
-	if err := r.cleanup(ctx, log, server); err != nil {
-		return ctrl.Result{}, err
+	// use update to not overwrite ServerMaintenanceRef if another maintenance was quicker
+	if err := r.Update(ctx, server); err != nil {
```

Comment:
Why not do a `PATCH` here?

Reply:
I want it to fail if in the meantime another ServerMaintenanceRef was set. (for whatever reason that may happen)